hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala

```diff
@@ -0,0 +1,388 @@
+  @transient private lazy val clients: Map[String, BlobServiceClient] = Map()
+
+  def getServiceClient(account: String): BlobServiceClient = {
+    clients.get(account) match {
+      case Some(client) => client
+      case None =>
+        val blobServiceClient = clientBuilder
+          .credential(credential)
+          .endpoint(s"https://$account.blob.core.windows.net")
+          .buildClient()
+        clients + (account -> blobServiceClient)
+        blobServiceClient
+    }
+  }
+}
+
+
+class AzureStorageFS(val serviceAccountKey: Option[String] = None) extends FS {
+  @transient private lazy val storageClientCache = serviceAccountKey match {
+    case None =>
+      val credential: DefaultAzureCredential = new DefaultAzureCredentialBuilder().build()
+      new AzureBlobServiceClientCache(credential)
+    case Some(keyData) =>
+      val kvs = JsonMethods.parse(keyData) match {
+        case json4s.JObject(values) => values.toMap
+      }
+
+      val appId = kvs("appId").asInstanceOf[json4s.JString].s
+      val password = kvs("password").asInstanceOf[json4s.JString].s
+      val tenant = kvs("tenant").asInstanceOf[json4s.JString].s
+
+      val clientSecretCredential: ClientSecretCredential = new ClientSecretCredentialBuilder()
+        .clientId(appId)
+        .clientSecret(password)
+        .tenantId(tenant)
+        .build();
+      new AzureBlobServiceClientCache(clientSecretCredential)
+  }
+
+  def getBlobServiceClient(account: String): BlobServiceClient = {
+    storageClientCache.getServiceClient(account)
+  }
+
+  def getBlobClient(filename: String): BlobClient = {
+    val (account, container, path) = getAccountContainerPath(filename)
+    getBlobServiceClient(account).getBlobContainerClient(container).getBlobClient(path)
+  }
+
+  def getContainerClient(filename: String): BlobContainerClient = {
+    val (account, container, _) = getAccountContainerPath(filename)
+    getBlobServiceClient(account).getBlobContainerClient(container)
+  }
+
+  def openNoCompression(filename: String): SeekableDataInputStream = {
+    val blobClient: BlobClient = getBlobClient(filename)
+
+    val is: SeekableInputStream = new FSSeekableInputStream {
+      private[this] var client: BlobClient = blobClient
+
+      def fill(): Unit = {
+        bb.clear()
+
+        // read some bytes
+        val outputStream: ByteArrayOutputStream = new ByteArrayOutputStream()
+        val blobProperties = client.getProperties
+        val blobSize = blobProperties.getBlobSize
+
+        val count = Math.min(blobSize - pos, bb.remaining() - 1)
+        if (count <= 0) {
+          eof = true
+          return
+        }
+
+        // calculate the minimum of the bytes remaining in the buffer and the bytes remaining
+        // to be read in the blob
+        client.downloadWithResponse(
+          outputStream, new BlobRange(pos, count),
+          null, null, false, Duration.ofMinutes(1), null)
```

Comment:
I don't think I can. I tried adding the parameter names and IntelliJ didn't like that. I think since the actual library is written in Java which doesn't have named arguments, it might not be possible to add them in